### PR TITLE
Retrieve NonClassEvents from database

### PR DESCRIPTION
**Pull request by rmainseas**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality
      to change)

## Checklist:

- [x] I have run `eslint` on the code
- [x] I have added JSDoc for all of my code (where applicable)
- [x] I have added tests to cover my changes.

## Priority:
- [x] Normal 
- [ ] High 

## Related Issues:

Delivers part of #210

### CODE REVIEW

**Comment by rmainseas** [comment]
@natalynyu FYI - you'll need the code in https://github.com/seas-computing/course-planner-etl/pull/40 in order to run the migrations for this

**Review by natalynyu — CHANGES_REQUESTED** [review]
This is looking good! It looks like there are a few places where there's a difference in the original `NonClassParent` (e.g. `contact`) vs. the new `NonClassParent` with the `contactName`, `contactEmail`, etc. 

For some reason I am getting a redis error when I try to bring up the app, and when I go to http://localhost:3001/course-planner/api/non-class-events, I'm getting this: 
image

Have you encountered this as well? I have the most recent code for this branch, tried to restart docker, and I'm on the VPN.

File: src/common/__tests__/data/nonClassEvents.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
+/**
+ * Data Science Reading Group
+ *
+ * Non-class parent for data science reading group formatted to match
+ * [[NonClassMeetingResponseDTO]].
+ */
+export const dataScienceReadingGroup = {
```
[the remark this answers is not in the document]
Comment (natalynyu) [reply src/common/__tests__/data/nonClassEvents.ts:7]: I'm seeing this error related to `contact`: 
image

File: src/server/nonClassEvent/NonClassParentView.entity.ts [hunk 2]
```diff
@@ -6,27 +6,49 @@ import {
...
 @ViewEntity('NonClassParentView', {
   expression: (connection: Connection):
   SelectQueryBuilder<NonClassParent> => connection.createQueryBuilder()
     .select('parent.id', 'id')
-    .addSelect('parent.contact', 'contact')
+    .addSelect('parent.contactName', 'contactName')
+    .addSelect('parent.contactEmail', 'contactEmail')
+    .addSelect('parent.contactPhone', 'contactPhone')
+    .addSelect('parent.notes', 'notes')
+    .addSelect('parent.expectedSize', 'expectedSize')
```
[the remark this answers is not in the document]
Comment (rmainseas) [reply src/server/nonClassEvent/NonClassParentView.entity.ts:21]: In instances like this where TypeORM is expecting a column name, it actually auto quotes the properties for you. For example, this query generated the following SQL:

```sql
SELECT "p"."id" AS "p_id",
  "p"."contactName" AS "p_contactName",
  "p"."contactEmail" AS "p_contactEmail",
  "p"."contactPhone" AS "p_contactPhone",
  "p"."notes" AS "p_notes",
  "p"."title" AS "p_title",
  "p"."expectedSize" AS "p_expectedSize",
  "p"."area" AS "p_area",
  "spring"."id" AS "spring_id",
  "spring"."academicYear" AS "spring_academicYear",
  "spring"."term" AS "spring_term",
  "spring"."nonClassParentId" AS "spring_nonClassParentId",
  "spring_meetings"."id" AS "spring_meetings_id",
  "spring_meetings"."day" AS "spring_meetings_day",
  "spring_meetings"."startTime" AS "spring_meetings_startTime",
  "spring_meetings"."endTime" AS "spring_meetings_endTime",
  "spring_meetings_room"."id" AS "spring_meetings_room_id",
  "spring_meetings_room"."name" AS "spring_meetings_room_name",
  "spring_meetings_room"."campus" AS "spring_meetings_room_campus",
  "spring_meetings_room"."capacity" AS "spring_meetings_room_capacity",
  "fall"."id" AS "fall_id",
  "fall"."academicYear" AS "fall_academicYear",
  "fall"."term" AS "fall_term",
  "fall"."nonClassParentId" AS "fall_nonClassParentId",
  "fall_meetings"."id" AS "fall_meetings_id",
  "fall_meetings"."day" AS "fall_meetings_day",
  "fall_meetings"."startTime" AS "fall_meetings_startTime",
  "fall_meetings"."endTime" AS "fall_meetings_endTime",
  "fall_meetings_room"."id" AS "fall_meetings_room_id",
  "fall_meetings_room"."name" AS "fall_meetings_room_name",
  "fall_meetings_room"."campus" AS "fall_meetings_room_campus",
  "fall_meetings_room"."capacity" AS "fall_meetings_room_capacity"
FROM "NonClassParentView" "p"
  LEFT JOIN "NonClassEventView" "spring" ON spring."nonClassParentId" = p."id"
  AND "spring"."term" = $1
  AND spring."academicYear" = $2
  LEFT JOIN "MeetingListingView" "spring_meetings" ON spring_meetings."nonClassEventId" = "spring"."id"
  LEFT JOIN "RoomListingView" "spring_meetings_room" ON "spring_meetings_room"."id" = "spring_meetings"."roomId"
  LEFT JOIN "NonClassEventView" "fall" ON fall."nonClassParentId" = p."id"
  AND "fall"."term" = $3
  AND fall."academicYear" = $4
  LEFT JOIN "MeetingListingView" "fall_meetings" ON fall_meetings."nonClassEventId" = "fall"."id"
  LEFT JOIN "RoomListingView" "fall_meetings_room" ON "fall_meetings_room"."id" = "fall_meetings"."roomId"
ORDER BY "p"."area" ASC
```

File: src/server/migrations/1620838443633-UpdateNonClassParentModel.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+/**
+ * This migration adds additional fields to the [[NonclassParent]] model such
```
[the remark this answers is not in the document]
Comment (rmainseas) [reply src/server/migrations/1620838443633-UpdateNonClassParentModel.ts]: Good catch. Fixed in 1d96e92f72ec983304f40822c567e3d9cd07db47

File: tests/mocks/database/population/data/index.ts [hunk 4]
```diff
@@ -74,7 +74,10 @@ export interface SemesterData {
 
 export interface NonClassMeetingData {
   title: string;
-  contact: string;
+  contactName: string;
+  contactEmail: string;
+  contactPhone: string;
+  expectedSize: number;
```
[the remark this answers is not in the document]
Comment (rmainseas) [reply tests/mocks/database/population/data/index.ts]: Good catch. Fixed in 821164d21821c50ee675842b6ff142db44b54f80

File: tests/mocks/database/population/data/nonClassMeetings.ts [hunk 5]
```diff
@@ -3,7 +3,10 @@ import { DAY } from '../../../../../src/common/constants';
 export const nonClassMeetings = [
```
[the remark this answers is not in the document]
Comment (rmainseas) [reply tests/mocks/database/population/data/nonClassMeetings.ts:3]: I agree. Done in f71687aaf0ea778c71ed882e278982a939fb474d

File: tests/mocks/database/population/nonclassevents.population.ts [hunk 6]
```diff
@@ -44,11 +48,17 @@ export class NonClassEventPopulationService
...
     return this.parentRepository.save(
       nonClassMeetings.map((parentData) => {
         const nonClassParent = new NonClassParent();
         nonClassParent.title = parentData.title;
-        nonClassParent.contact = parentData.contact;
+        nonClassParent.contactName = parentData.contactName;
+        nonClassParent.contactEmail = parentData.contactEmail;
+        nonClassParent.contactPhone = parentData.contactPhone;
+        nonClassParent.expectedSize = parentData.expectedSize;
+        nonClassParent.area = area;
```
[the remark this answers is not in the document]
Comment (rmainseas) [reply tests/mocks/database/population/nonclassevents.population.ts]: Yes, it should. Fixed in https://github.com/seas-computing/course-planner/pull/232/commits/f71687aaf0ea778c71ed882e278982a939fb474d

File: tests/integration/server/nonclassevents/nonclassevent.service.test.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    describe('Meetings', function () {
+      let dbMeetings: Meeting[];
+      let meetingRepository: Repository<Meeting>;
+      beforeEach(async function () {
+        meetingRepository = testModule.get(getRepositoryToken(Meeting));
+        dbMeetings = await meetingRepository.find({
+          relations: ['room', 'room.building'],
+        });
+      });
+      it('Should format the startTimes and endTimes as HH:MM AM', async function () {
```
[the remark this answers is not in the document]
Comment (rmainseas) [reply tests/integration/server/nonclassevents/nonclassevent.service.test.ts]: Yes, it should. It should also probably match the actual date format I'm using in the test case too :upside_down_face: . Well spotted. Fixed in 19d0b3db3fd7935593e206b30e4d62cddd51de7f

File: src/common/__tests__/data/nonClassEvents.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmainseas) [reply src/common/__tests__/data/nonClassEvents.ts:7]: Ah, I think I missed an instance of the test data not quite being in sync with the new schema. Weirdly, I never got those errors on my end when I ran the tests...

Either way - I fixed that in 242a1d64684c848a2cebde08da0c49c7c99b5586

File: tests/mocks/database/population/__tests__/population.service.test.ts [hunk 8]
```diff
@@ -268,10 +268,15 @@ describe('Population Service', function () {
...
+          const {
+            contactName, contactPhone, contactEmail, title, expectedSize,
+          } = nonClassEvent.nonClassParent;
           const testMeeting = testData.nonClassMeetings.find((nonClass) => (
             nonClass.title === title
-            && nonClass.contact === contact
+            && nonClass.contactName === contactName
+            && nonClass.contactPhone === contactPhone
+            && nonClass.contactEmail === contactEmail
+            && nonClass.expectedSize === expectedSize
```
Comment (natalynyu) [inline tests/mocks/database/population/__tests__/population.service.test.ts:279]: Should we include the `notes` here too?

File: tests/integration/server/nonclassevents/nonclassevent.controller.test.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        const response = await request(api).get('/api/non-class-events');
+
+        strictEqual(response.ok, false);
+        strictEqual(response.status, HttpStatus.FORBIDDEN);
+        strictEqual(mockNonClassEventService.find.callCount, 0);
+      });
+    });
+    describe('User is authenticated', function () {
+      it('it returns all the non-class events in the database', async function () {
+        authStub.resolves();
```
Comment (natalynyu) [inline tests/integration/server/nonclassevents/nonclassevent.controller.test.ts:88]: Should this page also only be accessible by admin users only?

**Review by natalynyu — CHANGES_REQUESTED** [review]
This is looking great! I just had a few last comments. Also I was wondering if this `nonClassParentId` inside of `spring` and `fall` are necessary, since it's already in the top level of the object? Seems like the request would just have the `id` property and not the additional two `nonClassParentId` properties, since they're each the same value?
image

File: tests/integration/server/nonclassevents/nonclassevent.controller.test.ts (hunk 9 above)
Comment (rmainseas) [reply tests/integration/server/nonclassevents/nonclassevent.controller.test.ts:88]: Interesting question. I actually don't know. Add/Edit, certainly. View is a little less clear. I'll ask Chris.

File: tests/mocks/database/population/__tests__/population.service.test.ts (hunk 8 above)
Comment (rmainseas) [reply tests/mocks/database/population/__tests__/population.service.test.ts]: D'oh. Yes, we should. Fixed in 320a255f44fbc2d91ba58c54439a1efc3c81b062

**Comment by rmainseas** [comment]
@natalynyu `nonClassParentId` isn't _needed_  in the data by the client, but it's needed in the query in order to have the `join` work correctly. Though, if you know of a way to do it without, I'm happy to change it.

File: tests/integration/server/nonclassevents/nonclassevent.controller.test.ts (hunk 9 above)
Comment (rmainseas) [reply tests/integration/server/nonclassevents/nonclassevent.controller.test.ts:88]: I just spoke to Chris about this and he said that all users should be able to _see_ the data. There will be some additional restrictions involved in adding/editing data - but that's another ticket.

As far as this ticket is concerned - all authenticated users should be able to see this data.

**Review by natalynyu — COMMENTED** [review]
@rmainseas I found that when I remove this @ViewColumn from the `NonClassEventView`:

image

...it removes the `NonClassParentId`:
image

**Comment by rmainseas** [comment]
Ah right, interesting. I had thought that removing `@ViewColumn()` from an entity would remove that column from the actual database view. I guess that's controlled by SQL query at the top of the view entity itself. For some reason I'd forgotten that.

Either way, that's been updated in b27f011cebeff6204444681a64f2f257d876eff8 - good call.

**Review by natalynyu — APPROVED** [review]
Approved! Looks great!

**Pull request merged**